### PR TITLE
Anonymous / No-auth mode

### PR DESCRIPTION
**Pull request by foot**
**What changed?**

Introduces a cli-flag to enable "anonymous" mode.

This allows the dashboard to be configured without auth, no login screen will be shown. Its designed to be used with external authentication systems like auth-proxies.

The kubernetes `User` that is impersonated can be configured via the flag value.

**Why was this change made?**

Allows external auth systems to be used instead of built in OIDC.

**How was this change implemented?**

- A new principal was added

**How did you validate the change?**

- Unit tests
- Manual testing

**Documentation Changes**

Yes, included.

### CODE REVIEW

File: cmd/gitops-server/cmd/cmd.go [hunk 1]
```diff
@@ -136,6 +141,10 @@ func runCmd(cmd *cobra.Command, args []string) error {
 
 	featureflags.SetFromEnv(os.Environ())
 
+	if cmd.Flags().Changed(InsecureNoAuthenticationUserFlag) && options.NoAuthUser == "" {
+		return fmt.Errorf("%s flag set but no user specified", InsecureNoAuthenticationUserFlag)
+	}
```
Comment (foot) [inline cmd/gitops-server/cmd/cmd.go:146]: Is there a more idiomatic way to do this validation with cobra?

File: pkg/server/auth/auth_method_type.go [hunk 2]
```diff
@@ -17,13 +17,27 @@ const (
...
+	// Anonymous
+	Anonymous
 )
 
 // This is a function to mimic a const slice
 func DefaultAuthMethods() []AuthMethod {
 	return []AuthMethod{UserAccount, OIDC}
 }
 
+func AllUserAuthMethods() []string {
```
Comment (foot) [inline pkg/server/auth/auth_method_type.go:30]: Docstring

File: pkg/server/auth/init.go [hunk 3]
```diff
@@ -13,16 +13,16 @@ import (
...
 
 	authMethods, err := ParseAuthMethodArray(authMethodStrings)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(authMethods) == 0 {
-		return nil, fmt.Errorf("no authentication methods set")
+	if noAuthUser != "" {
+		authMethods = map[AuthMethod]bool{Anonymous: true}
```
Comment (foot) [inline pkg/server/auth/init.go:25]: Ooops, restore this check

File: pkg/server/auth/auth.go [hunk 4]
```diff
@@ -188,14 +188,17 @@ func WithAPIAuth(next http.Handler, srv *AuthServer, publicRoutes []string) http
...
 				adminAuth := NewJWTAdminCookiePrincipalGetter(srv.Log, srv.tokenSignerVerifier, IDTokenCookieName)
 				multi.Getters = append(multi.Getters, adminAuth)
 			}
 
 		case TokenPassthrough:
 			tokenAuth := NewBearerTokenPassthroughPrincipalGetter(srv.Log, nil, AuthorizationTokenHeaderName, srv.kubernetesClient)
 			multi.Getters = append(multi.Getters, tokenAuth)
+
+		case Anonymous:
+			multi.Getters = []PrincipalGetter{NewAnonymousPrincipalGetter(srv.Log, srv.noAuthUser)}
```
Comment (foot) [inline pkg/server/auth/auth.go:201]: Should this be wrapped in a `if featureflags.IsSet(FeatureFlagAnon) {` kind of thing?

Comment (foot) [reply pkg/server/auth/auth.go]: Wrapped it, we should never get there and this code could use some refactoring..

File: website/docs/configuration/anonymous-access.mdx [hunk 5]
```diff
@@ -0,0 +1,34 @@
+---
+title: Anonymous Access
+---
+
+:::danger Important
+Alone, this is an **insecure** method of securing your dashboard.
+
+It is designed to be used with other external authentication systems like auth proxies.
```
Comment (bigkevmcd) [inline website/docs/configuration/anonymous-access.mdx:8]: I'm not sure this bit is needed, it does nothing to support them.

**Review by bigkevmcd — CHANGES_REQUESTED** [review]

File: cmd/gitops-server/cmd/cmd.go [hunk 6]
```diff
@@ -177,7 +186,7 @@ func runCmd(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("couldn't get current namespace")
 	}
 
-	authServer, err := auth.InitAuthServer(cmd.Context(), log, rawClient, options.OIDC, options.OIDCSecret, namespace, options.AuthMethods)
+	authServer, err := auth.InitAuthServer(cmd.Context(), log, rawClient, options.OIDC, options.OIDCSecret, namespace, options.AuthMethods, options.NoAuthUser)
```
[the remark this answers is not in the document]
Comment (foot) [reply cmd/gitops-server/cmd/cmd.go:189]: Yeah good shout, it's a bit tangled up, we do have an AuthConfig, but that comes "later" and is generated inside of InitAuthServer somewhere. Will TAL

File: pkg/server/auth/init.go [hunk 7]
```diff
@@ -13,14 +13,18 @@ import (
 
 // InitAuthServer creates a new AuthServer and configures it for the correct
 // authentication methods.
-func InitAuthServer(ctx context.Context, log logr.Logger, rawKubernetesClient ctrlclient.Client, oidcConfig OIDCConfig, oidcSecret, namespace string, authMethodStrings []string) (*AuthServer, error) {
+func InitAuthServer(ctx context.Context, log logr.Logger, rawKubernetesClient ctrlclient.Client, oidcConfig OIDCConfig, oidcSecret, namespace string, authMethodStrings []string, noAuthUser string) (*AuthServer, error) {
 	log.V(logger.LogLevelDebug).Info("Registering authentication methods", "methods", authMethodStrings)
```
[the remark this answers is not in the document]
Comment (foot) [reply pkg/server/auth/init.go:17]: Yes! Good one, startup logging is so useful.

File: website/docs/configuration/anonymous-access.mdx (hunk 5 above)
Comment (foot) [reply website/docs/configuration/anonymous-access.mdx:8]: Yes, rephrase?

> If you decide to use anonymous access, you should be cautious about exposing the endpoint and consider implementing additional authentication measures such as an auth proxy.

Comment (bigkevmcd) [reply website/docs/configuration/anonymous-access.mdx:8]: Sounds better 👍🏻

**Comment by foot** [comment]
cc @LappleApple as you've been looking at the user-guide. This PR also adds some user-guide docs about this new anonymous access feature.

**Review by bigkevmcd — APPROVED** [review]

**Comment by bigkevmcd** [comment]
Thanks!

**Pull request merged**